README.rst
===============================
Song Comparison
===============================

.. image:: https://badge.fury.io/py/song_comparison.png
    :target: http://badge.fury.io/py/song_comparison
    
.. image:: https://travis-ci.org/jw1123/song_comparison.png?branch=master
        :target: https://travis-ci.org/jw1123/song_comparison

.. image:: https://pypip.in/d/song_comparison/badge.png
        :target: https://crate.io/packages/song_comparison?version=latest


* This is a script to help extract audio features and to calculate distances between songs with choice parameters.
* All data is stored in a database of MongoDB, so you should download MongoDB.
* The script only works with wave files, being named in a way described later. You can give mp3 or m4a files (ID3 taged!) 
* to the script and it will convert them to wave files with ffmpeg for you. Make sure ffmpeg (file conversion) and mutagen (ID3 tag) are installed.
* For audio feature extraction, this script is using the bregman audio toolbox and aubio toolbox (for the rhythm) 
* which you have to download and install (make sure both work).
* There is also a set of packages for python: you should have installed numpy, matplotlib, pymongo

Prerequisites:

- python 2.7
- bregman
- aubio
- mongoDB
- ffmpeg
- mutagen
- numpy
- pymongo



Note:

To install aubio, download or clone the latest version on https://github.com/piem/aubio. If you are on a OS X 10.9, then open the wscript file and comment line 86 and 87 out. Follow the instructions from the README. If ./waf build does not work, then reconfigure using this command line: ./waf configure —-disable-sndfile —-disable-samplerate (for some reason, these packages are not necessary). Now it should build and install properly. As we are interested in the python bindings, go to the python folder and follow the instructions in the README. To install it type sudo python setup.py install after building is completed.


Features:
---------------------------------------------------------------------------------------

There is a list of features you can chose from to be extracted from your audio files.

Chromagram
HighQuefrencyChromagram
HighQuefrencyLogFrequencySpectrum
HighQuefrencyMelSpectrum
LinearFrequencySpectrum
LinearFrequencySpectrumCentroid
LinearFrequencySpectrumSpread
LinearPower
LogFrequencySpectrum
LogFrequencySpectrumCentroid
LogFrequencySpectrumSpread
LowQuefrencyLogFrequencySpectrum
LowQuefrencyMelSpectrum
MelFrequencyCepstrum (MFCC)
MelFrequencySpectrumCentroid
MelFrequencySpectrumSpread
RMS
dBPower
BPM


Paramaters:
---------------------------------------------------------------------------------------

For every feature, you can choose to set the following parameters to your liking. If you
don't set the values yourself, they will be set to the default values listed here.

'sample_rate': 44100, # The audio sample rate
'nbpo': 12,           # Number of Bands Per Octave for front-end filterbank
'ncoef' : 10,         # Number of cepstral coefficients to use for cepstral features
'lcoef' : 1,          # Starting cepstral coefficient
'lo': 62.5,           # Lowest band edge frequency of filterbank
'hi': 16000,          # Highest band edge frequency of filterbank
'nfft': 16384,        # FFT length for filterbank
'wfft': 8192,         # FFT signal window length
'nhop': 4410,         # FFT hop size
'log10': False,       # Whether to use log output
'magnitude': True,    # Whether to use magnitude (False=power)
'intensify' : False,  # Whether to use critical band masking in chroma extraction
'onsets' : False,     # Whether to use onset-synchronus features
'verbosity' : 1       # How much to tell the user about extraction




Usage of document type toolbox:

The document toolbox uses a configuration file to retrieve the features you want to extract and the parameters to each feature. Edit the configuration file likewise to this model (if you want the default parameters values, just write default=0):

[Feature]
parameter1=x
parameter2=y
etc.

Example:

[MelFrequencyCepstrum]
sample_rate=22000
ncoef=5



Type the following command into your terminal to use the document toolbox: python main.py -(-)option (argument) (path)
The following options are available:
-c, --conversion		    : conversion of mp4 and mp3 files to wave files
-e, --extraction            : feature extraction of either a new set or adding features to a set
-d, --distance              : calculate the distance between each song
-g, --graphdist             : creates a graphml file with song nodes and distance edges
NOTE: If you use conversion or extraction, you have to add the path to the file with your mp4/mp3 files or wave files respectively
EXAMPLE: python main.py -e /Users/myname/Document/wave/




* Free software: BSD license
* Documentation: http://song_comparison.rtfd.org.

Features
--------

* TODO

HISTORY.rst
.. :changelog:

History
-------

0.1.0 (26.11.2013)
++++++++++++++++++

* First release on PyPI.

tests/test_song_comparison.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_song_comparison
----------------------------------

Tests for `song_comparison` module.
"""

import unittest

from song_comparison import song_comparison


class TestSong_comparison(unittest.TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()

song_comparison/main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


from sys import argv
from timeit import timeit
from getopt import getopt
from subprocess import call
from pymongo import Connection
from time import sleep

# print timeit("Distance()", setup="from distance import Distance", number=2)

class Main():
    def __init__(self):
        try:
            con = Connection()
        except:
            call(["osascript", "-e", 'tell app "Terminal" to do script "mongod"'])
            sleep(3)
            con = Connection()
        #__________________________________________________________________________
        # Change the name of the database and collections to your liking
        db = con.feature_extraction_database                    # #####REPLACE#####
        song_features = db.song_features_collection             # #####REPLACE#####
        distance_features = db.distance_features_collection     # #####REPLACE#####
        #__________________________________________________________________________

        options, rem = getopt(argv[1:], 'c:e:d:g:h', ['conversion=',
            'extraction=', 'distance', 'graphdist', 'help'])

        for opt, arg in options:
            if opt in ('-c', '--conversion'):
                from conversion import Conversion
                c = Conversion(arg)
            elif opt in ('-e', '--extraction'):
                from extraction import Extraction
                e = Extraction(arg, song_features)
            elif opt in ('-d', '--distance'):
                from distance import Distance
                d = Distance(song_features, distance_features)
            elif opt in ('-g', '--graphdist'):
                from graphdist import Graphdist
                g = Graphdist(song_features, distance_features)
            elif opt in ('-h', '--help'):
                print """The following options are available:
                -c, --conversion mp3/m4a
                =>  conversion of mp4 and mp3 files to wave files

                -e, --extraction
                =>  feature extraction of either a new set or
                    adding features to a set

                -d, --distance
                =>  calculate the distance between each song

                -g, --graphdist
                => creates a graphml file with song nodes and distance edge

                NOTE: If you use conversion or extraction, you have to add
                the path to the file with your mp4/mp3 filesor wave files
                respectively
                EXAMPLE: python main.py -e /Users/myname/Document/wave/
                """

if __name__ == "__main__":
    m = Main()



song_comparison/distance.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from numpy import linalg, array, arange, sqrt, concatenate
from time import sleep


class Distance:

    def __init__(self, song_collection, distance_collection):

        print "DISTANCE CALCULATION"
        #___________________________________________________
        # Change the coefficients for the features you chose
        coeff = [0.3, 0.3, 0.1, 0.2]   # ###REPLACE####
        #___________________________________________________
        self.so = song_collection
        self.di = distance_collection

        self.iterating(coeff)

        # To iterate and show all collections
        # of the distance database
        for d in self.di.find():
            print d

    def iterating(self, coeff):
        s = self.so.find(timeout=False)
        r = self.so.find(timeout=False)
        ski = 1
        j = 1
        for song1 in s:
            # Compare a song only to other songs that
            # have a higher song_id, thus skip the
            # songs with a lower id
            r.skip(ski)
            for song2 in r:
                # Check if both songs have the same number of features,
                # if not, they cannot be compared
                if song1['features'].keys().sort() == song2['features'].keys().sort():
                    # Calling the function that runs through all the features
                    # and calls the distance calculation function
                    distance_dict, weight = self.browse_features(song1, song2, coeff)
                    # Only consider pairs that lie beneath a certain threshold
                    # 0.00987 is the median of my last run
                    if weight < 0.00987:
                        # Check if the collection already exists, if not,
                        # insert the dictionary
                        if type(self.di.find_one({"source": song1['song_id']})) != dict or\
                         type(self.di.find_one({"source": song2['song_id']})) != dict:
                            self.insert_dictionary(song1, song2, distance_dict, weight)
                        # Check which song is the source and which song is the target
                        elif self.di.find_one({"source": song1['song_id']})['target'] == song2["song_id"]:
                            self.set_dictionary(song1, song2, distance_dict, weight)
                        elif self.di.find_one({"source": song2['song_id']})['target'] == song1["song_id"]:
                            self.set_dictionary(song2, song1, distance_dict, weight)
                        print j
                        j += 1
                else:
                    print "Songs need to have the same features"
                    print song1['metadata']['title'], ": ", song1['features']
                    print song2['metadata']['title'], ": ", song2['features']
                    break
            ski += 1
            r.rewind()
        print "Distance calculation successfully completed!"

    def insert_dictionary(self, song1, song2, distance_dict, weight):
        # Creating a dictionary that is inserted in
        # the distance database
        distance_insert_dictionary = {
            'source': song1['song_id'],
            'target': song2['song_id'],
            'feature_distance': distance_dict,
            'weight': weight}
        self.di.insert(distance_insert_dictionary)

    def set_dictionary(self, song1, song2, distance_dict, weight):
        # For an existing collection, a query dictionary
        # allows to find the right collection and the
        # set dictionary replaces the existing distance
        # dictionary
        query_dictionary = {
            'source': song1['song_id'],
            'target': song2['song_id']}
        set_dictionary = {
            'feature_distance': distance_dict,
            'weight': weight}
        self.di.update(query_dictionary, {"$set": set_dictionary})

    def browse_features(self, song1, song2, coeff):
        distance_dict = {}
        summe, dd, ddd, ii = 0, 0, 0, 0
        l = max(song1['metadata']['length'],
                song2['metadata']['length'])
        # Browsing through all features
        for i in song1['features']:
            jii = 0
            # Browsing through all parameter names
            for k in arange(len(song1['features'][i])):
                for h in song1['features'][i][k]:
                    a = song1['features'][i][k][h]['data']
                    b = song2['features'][i][k][h]['data']
                    # Checking the type of features a and b
                    # If float => bpm_comparison
                    # If not => array comparison
                    if type(a) == float and type(b) == float:
                        dd = self.bpm_comparison(a, b)/l
                    else:
                        dd = self.distance_calculation(a, b)/l
                    # Assign the correct coefficient to the
                    # matching feature and multiply the
                    # calculated distance with it
                    if i == 'MelFrequencyCepstrum':
                        ddd = coeff[0]*dd
                    elif i == 'LinearPower':
                        ddd = coeff[1]*dd
                    elif i == 'BPM':
                        ddd = coeff[2]*dd
                    elif i == 'Chromagram':
                        ddd = coeff[3]*dd
                    else:
                        ddd = dd
                    # Check if the feature already exists in the dictionary,
                    # if not (jii == 0) then create the field, if yes then
                    # append the parameter dictionary to the existing list
                    if jii == 0:
                        distance_dict.update({i: [{h: dd}]})
                    else:
                        distance_dict[i].append({h: dd})

                    jii += 1
                    summe += ddd
                    ii += 1
        # Return the distance dictionary and the weight
        return distance_dict, summe/ii

    def bpm_comparison(self, x, y):
        # Compare two bpm's, if they are close to each
        # other or close to the multiple of each other,
        # then the bpm distance is reduced by a factor
        # of 4
        if x % y < 6:
            return abs(x-y)*0.25
        else:
            return abs(x-y)

    def distance_calculation(self, x, y):
        # Determine which list is longer and invoking the reshaping function
        # to even their length (otherwise we cannot compare their values)
        p = array(x)
        q = array(y)

        if p.shape > q.shape:
            d = self.reshape(p, q)
            return d
        elif p.shape < q.shape:
            d = self.reshape(q, p)
            return d
        elif p.shape == q.shape:
            d = linalg.norm(p-q)
            return d
        else:
            return 0

    def reshape(self, b, a):
        # Reshaping and filling of the smaller array
        template = array([])
        new_array = []

        # Check if both arrays are unidimensional or not
        # If the first dimension of the arrays are unequal,
        # we have one dimensional arrays
        # This is due to the Bregman extraction tool, which
        # has an already set number of rows for two dimensional
        # arrays (ex. all MFCC have 95 rows)
        if b.shape[0] == a.shape[0]:
            x = b.shape[1]
            l = int(x/a.shape[1])
            # The smaller array has an l-times less columns
            # Run through all rows (for i in a)
            for i in a:
                template = i
                # "Multiply" l times the columns
                for j in arange(l-1):
                    template = concatenate([template, i])
                # Fill up with the rest
                template = concatenate([template, i[:x-len(template)]])
                new_array.append(template)
        else:
            # Same principle as above, only with a one dimensional
            # array
            x = b.shape[0]
            l = int(x/a.shape[0])
            for j in arange(l):
                template = concatenate([template, a])
            template = concatenate([template, a[:x-len(template)]])
            new_array.append(template)

        return linalg.norm(b-array(new_array))


song_comparison/graphdist.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import networkx as nx
from matplotlib.pyplot import show
from os import path


class Graphdist:

    def __init__(self, song_collection, distance_collection):
        self.so = song_collection
        self.di = distance_collection
        self.savegraph()

    def savegraph(self):

        G = nx.DiGraph()
        # Iterate through all songs and create nodes
        # of them in the graph
        for s in self.so.find():
            G.add_node(s['song_id'], s['metadata'])
        # Iterate through all distances and create
        # edges between the songs of the graph
        for d in self.di.find():
            # Multiply the weight by 10000 because the values
            # are rounded up in Gephi
            G.add_edge(d['source'], d['target'], weight=10000*d['weight'])
        # Create the GraphML file
        nx.write_graphml(G, path.dirname(path.realpath("graph.py"))
                                        + '/song_comparison.graphml')


song_comparison/conversion.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from os import walk
from mutagen.id3 import ID3
from mutagen.mp4 import MP4
from subprocess import call, STDOUT, PIPE


class Conversion():

    def __init__(self, path):

        print "CONVERSION"

        # Create a folder for the wave files
        input_path = path
        call(["mkdir", path + "wave/"])
        output_path = path + "wave/"

        self.any_to_wave(input_path, output_path)

    def any_to_wave(self, inp, out):
        # Browse through all files in directory inp
        for root, dirs, files in walk(inp):
            for fil in files:
                end = fil[len(fil)-3:len(fil)]
                # Check if the file is an audio file
                if end in ["mp3", "mp4", "m4a"]:
                    if end == "mp3":
                        # Extract MP3 tags
                        tag = ID3(inp + fil)
                    elif end == "mp4" or end == "m4a":
                        # Extract MP4 and M4A tags
                        mp4 = MP4(inp + fil)
                        tag = mp4.MP4Tags

                    # Check for tags and assign a variable
                    if "TIT2" in tag:
                        title = str(tag["TIT2"])
                    else:
                        title = ''
                    if "TPE1" in tag:
                        artist = str(tag["TIT2"])
                    else:
                        artist = ''
                    if "TALB" in tag:
                        album = str(tag["TALB"])
                    else:
                        album = ''
                    if "TCON" in tag:
                        genre = str(tag["TCON"])
                    else:
                        genre = ''
                    if "TDRC" in tag:
                        year = str(tag["TDRC"])
                    else:
                        year = ''
                    # Call ffmpeg in the terminal and save the wave file
                    # with the tags as file name (if there is no tag,
                    # use old file name and with empty separators)
                    if title + artist + album + genre + year == '':
                        call(["ffmpeg", "-i", inp + fil, "-ar", "11000",
                            out + fil[:len(fil)-4] + "-*--*--*--*-" + ".wav"],
                            stderr=STDOUT, stdout=PIPE)
                    else:
                        call(["ffmpeg", "-i", inp + fil, "-ar", "11000",
                            out + title
                            + "-*-" + artist
                            + "-*-" + album
                            + "-*-" + genre
                            + "-*-" + year
                            + ".wav"],
                            stderr=STDOUT, stdout=PIPE)

        print "Conversion successfully completed!"


song_comparison/extraction.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from os import walk, path
from time import sleep
from bregman.suite import os, Chromagram, HighQuefrencyChromagram, \
    HighQuefrencyLogFrequencySpectrum, HighQuefrencyMelSpectrum, \
    LinearFrequencySpectrum, LinearFrequencySpectrumCentroid, \
    LinearFrequencySpectrumSpread, LinearPower, LogFrequencySpectrum, \
    LogFrequencySpectrumCentroid, LogFrequencySpectrumSpread,\
    LowQuefrencyLogFrequencySpectrum, LowQuefrencyMelSpectrum, \
    MelFrequencyCepstrum, MelFrequencySpectrumCentroid, \
    MelFrequencySpectrumSpread, RMS, dBPower
from contextlib import closing
from wave import open as wopen
from numpy import arange, median, mean
from numpy import sum as npsum
from aubio import tempo, source
from ConfigParser import ConfigParser


class Extraction:

    def __init__(self, path_wave, song_collection):
        print "EXTRACTION"
        # List of all the features that can
        # be extracted so that they can be used as functions
        self.features_list = {'Chromagram': Chromagram,
        'HighQuefrencyChromagram': HighQuefrencyChromagram,
        'HighQuefrencyLogFrequencySpectrum': HighQuefrencyLogFrequencySpectrum,
        'HighQuefrencyMelSpectrum': HighQuefrencyMelSpectrum,
        'LinearFrequencySpectrum': LinearFrequencySpectrum,
        'LinearFrequencySpectrumCentroid': LinearFrequencySpectrumCentroid,
        'LinearFrequencySpectrumSpread': LinearFrequencySpectrumSpread,
        'LinearPower': LinearPower, 'LogFrequencySpectrum': LogFrequencySpectrum,
        'LogFrequencySpectrumCentroid': LogFrequencySpectrumCentroid,
        'LogFrequencySpectrumSpread': LogFrequencySpectrumSpread,
        'LowQuefrencyLogFrequencySpectrum': LowQuefrencyLogFrequencySpectrum,
        'LowQuefrencyMelSpectrum': LowQuefrencyMelSpectrum,
        'MelFrequencyCepstrum': MelFrequencyCepstrum,
        'MelFrequencySpectrumCentroid': MelFrequencySpectrumCentroid,
        'MelFrequencySpectrumSpread': MelFrequencySpectrumSpread,
        'RMS': RMS, 'dBPower': dBPower, 'BPM': 'BPM'}

        #__________________________________________________________
        path_data = path.dirname(path.realpath("configure.txt")) + "/configure.txt"
        #_________________________________________________________________

        config = ConfigParser()
        config.read(path_data)
        feat_dic = {}
        # Storing the parameter values and features from the
        # configuration file in a dictionary
        for feat in config.sections():
            i = 0
            for param in config.options(feat):
                if i == 0:
                    feat_dic.update({feat: {param: int(config.get(feat, param))}})
                else:
                    feat_dic[feat].update({param: int(config.get(feat, param))})
                i += 1

        mdbfeat = {}

        for feat in feat_dic:
            a = None
            b = ''
            # Convert parameters into a name,
            # to classify the extracted features
            if len(feat_dic[feat]) > 1:
                for i in feat_dic[feat]:
                    b += self.parameter_short({i: feat_dic[feat][i]})
                mdbfeat.update({feat: b})
            else:
                mdbfeat.update({feat: self.parameter_short(feat_dic[feat])})

        self.so = song_collection
        self.par_short = mdbfeat
        self.features = feat_dic

        self.extract(path_wave)

        # To iterate through the song database
        # and show all the song collections
        for s in self.so.find():
            print s

    def parameter_short(self, a):
        b = ''
        # Check for all the parameters and
        # combine them as abreviations
        for p in a:
            if p == "default" or 0:
                b = "default"
            else:
                if p == "sample_rate":
                    b += "s" + str(a[p])[:len(str(a[p]))-3]
                elif p == "nbpo":
                    b += "nb" + str(a[p])
                elif p == "ncoef":
                    b += "nc" + str(a[p])
                elif p == "lcoef":
                    b += "lc" + str(a[p])
                elif p == "lo":
                    b += "lo" + str(a[p])
                elif p == "hi":
                    b += "hi" + str(a[p])[:len(str(a[p]))-3]
                elif p == "nfft":
                    b += "nf" + str(a[p])
                elif p == "wfft":
                    b += "w" + str(a[p])
                elif p == "nhop":
                    b += "nh" + str(a[p])
                elif p == "log10":
                    b += "l10" + str(a[p])[0]
                elif p == "magnitude":
                    b += "m" + str(a[p])[0]
                elif p == "intensity":
                    b += "i" + str(a[p])[0]
                elif p == "onsets":
                    b += "o" + str(a[p])[0]
                else:
                    b += "v" + str(a[p])
        return b

    def BPM(self, path, param):
        """ Calculate the beats per minute (bpm) as a rhythm feature.
            path: path to the file
            param: dictionary of parameters
        """
        try:
            win_s = param['wfft']
            samplerate = param['sample_rate']
        except:
            win_s = 512                 # fft size
            samplerate = 11000

        hop_s = win_s / 2           # hop size

        s = source(path, samplerate, hop_s)
        samplerate = s.samplerate
        o = tempo("default", win_s, hop_s, samplerate)

        # Tempo detection delay, in samples
        # default to 4 blocks delay to catch up with
        delay = 4. * hop_s

        # List of beats, in samples
        beats = []

        # Total number of frames read
        total_frames = 0
        while True:
            samples, read = s()
            is_beat = o(samples)
            if is_beat:
                this_beat = int(total_frames - delay + is_beat[0] * hop_s)
                beats.append(this_beat)
            total_frames += read
            if read < hop_s:
                break

        # Convert samples to seconds
        beats = map(lambda x: x/float(samplerate), beats)

        bpms = [60./(b-a) for a, b in zip(beats[:-1], beats[1:])]

        if samplerate == 11000:
            b = median(bpms)*4
        elif samplerate == 22000:
            b = median(bpms)*2
        else:
            b = median(bpms)
        return b

    def sliding_window(self, x, coeff):
        """ The principle of the sliding window is the following:
            We have an array that we want to reduce by a certain factor
            (16 by default). We only reduce the array in the column
            dimension, as the row dimension is critical for certain
            features (as the 12 step chromagram). So we take the array
            and average over coeff + coeff/2 cases (50 percent overlapping)
            and store it as a new array.
            Example: if the initial shape of x is (50,1600), after the
            sliding window has been applied, its new shape is (50,100)
        """
        new_array = []
        # Determine the dimensions of x
        if x.shape != (x.shape[0],):
            for i in x:
                template = []
                for j in arange(int(len(i)/coeff)):
                    if(j != 0):
                        # If it is neither the first, nor the last case
                        template.append(float(npsum(i[j*coeff-coeff/2:(j+1)*coeff])/(coeff+coeff/2)))
                    elif j == int(len(i)/coeff)-1:
                        # Last case, filling up
                        template.append(float(npsum(i[j*coeff-coeff/2:])/len(i[j*coeff-coeff/2:])))
                    else:
                        # First case
                        template.append(float(npsum(i[:(j+1)*coeff])/coeff))
                new_array.append(template)
            xshape = (len(new_array),len(new_array[0]))
        else:
            template = []
            for j in arange(int(x.shape[0]/coeff)):
                if(j != 0):
                    template.append(float(npsum(x[j*coeff-coeff/2:(j+1)*coeff])/(coeff+coeff/2)))
                elif j == int(x.shape[0]/coeff)-1:
                    # Last case, filling up
                    template.append(float(npsum(x[j*coeff-coeff/2:])/len(x[j*coeff-coeff/2:])))
                else:
                    # First case
                    template.append(float(npsum(x[:(j+1)*coeff])/coeff))
            new_array = template
            xshape = (len(new_array),1)

        return new_array, xshape

    def extract_features(self, x):
        """ Extract all features and reducing them invoking the wind function.
            x: path to the file
            para: list of parameter dictionaries
        """
        fea_dic = {}

        # Recognize the features and launch their respective functions
        for f in self.features:
            if f != 'BPM':
                if self.features[f] != {"default": 0}:
                    # **para[i] gets the parameter dictionary and uses
                    # it as an argument for the bregman functions
                    fea_dic.update({f: self.features_list[f](x, **self.features[f])})
                else:
                    # Extraction with default values
                    fea_dic.update({f: self.features_list[f](x)})
            else:
                fea_dic.update({f: self.BPM(x, self.features[f])})

        fea_dic1 = {}
        # Convert the numpy arrays into binaries to store them in MongoDB
        # Change the sliding window reduction coefficient to your liking
        for fe in fea_dic:
            if 'Chromagram' in fe:
                d, s = self.sliding_window(fea_dic[fe].CHROMA, 16)
            elif fe == 'MelFrequencyCepstrum':
                d, s = self.sliding_window(fea_dic[fe].MFCC, 16)
            elif 'LinearFrequencySpectrum' in fe:
                d, s = self.sliding_window(fea_dic[fe].STFT, 16)
            elif fe == 'LinearPower' or fe == 'RMS' or fe == 'dBPower':
                d, s = self.sliding_window(fea_dic[fe].POWER, 16)
            elif fe == 'BPM':
                d = fea_dic[fe]
                s = (0, 0)
            else:
                d, s = self.sliding_window(fea_dic[fe].CQFT, 16)
            fea_dic1.update({fe: {"data": d, "row": s[0], "column": s[1],
                "parameters": self.features[fe]}})
        return fea_dic1



    def parameter_name(self, features_diction):
        # Combine the parameter name dictionary with the feature array
        # dictionary so as to have the final dictionary
        mdbfeat = {}
        for feat in self.par_short:
            mdbfeat.update({feat: [{self.par_short[feat]: features_diction[feat]}]})
        return mdbfeat

    def extract(self, direc):
        """ Iterating function, and inserting id, metadata and features to the collection.
            param: list of parameter dictionaries
        """
        i = 0

        for root, dirs, files in os.walk(direc):
            for file1 in files:
                if file1[len(file1)-3:] == "wav":
                    i += 1
                    w = wopen(direc+file1, "r")
                    # Exracting the file length
                    with closing(w) as f:
                        frame = f.getnframes()
                        rate = f.getframerate()

                    # Launch the actual extraction
                    dbfeat = self.parameter_name(self.extract_features(direc+file1))
                    # Split the file name in order to use all the metadata
                    meta = file1.split("-*-")

                    # Assign the metadata informations
                    if len(meta) == 5:
                        tit = meta[0]
                        art = meta[1]
                        alb = meta[2]
                        gen = meta[3]
                        yea = meta[4][:len(meta[4])-4]
                    elif len(meta) == 4:
                        tit = meta[0]
                        art = meta[1]
                        alb = meta[2]
                        gen = meta[3][:len(meta[3])-4]
                        yea = ''

                    if self.so.find({"metadata.title": tit}).count() == 0:
                        # If this particular song's collection does not
                        # exist, insert the dictionary
                        song_dictionary_insert = {
                                        'song_id': i,
                                        'metadata': {
                                        'title': tit,
                                        'artist': art,
                                        'album': alb,
                                        'genre': gen,
                                        'year': yea,
                                        'length': frame/float(rate)},
                                        'features': dbfeat}

                        self.so.insert(song_dictionary_insert)
                    else:
                        # If this particular song's collection already exists,
                        # then only update the feature dictionary
                        for k in dbfeat:
                            for par in dbfeat[k][0]:
                                a = "features."+k
                                self.so.update({"metadata.title": tit, "metadata.artist": art,
                                "metadata.album": alb}, {"$addToSet": {a: {par: dbfeat[k][0][par]}}})

                    w.close()
                    print i

        print "Extraction successfully completed!"
